<a href="https://colab.research.google.com/github/MarkNgendo/MTN-Cote-d-Ivoire-Project/blob/main/MTN_Cote_d'Ivoire_Product_Usage.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## PROBLEM STATEMENT


Currently MTN Cote d'Ivoire would like to upgrade its technology infrastructure for its mobile users in Ivory Coast. Studying the given dataset, how does MTN Cote d'Ivoire go about the upgrade of its infrastructure strategy within the given cities?

For analysis, the following questions were investigated:

Which cities have the highest user traffic based on provided data? Which MTN products are the most frequently used? Which cities bring in the highest profits per product based on provided data? Which are the busiest hours of the day in terms of user traffic?

## DATA RELEVANCE
Below is a list of datasets that we shall be working on:


*  cells_geo_description.xlsx
*   cells_geo.csv 
*   CDR_description.xlsx 
*   CDR 20120507 
*   CDR 20120508 
*   CDR 20120509 










## Importing Libraries





In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

## Loading Datasets

In [2]:
cells = pd.read_csv('cells_geo.csv', delimiter=';')
cells.head()

Unnamed: 0   VILLES      STATUS  ... AREA     CELL_ID   SITE_CODE
0           0   ADJAME  In Service  ...  NaN  fbb2e75af3  0bb492ef17
1           1   ADJAME  In Service  ...  NaN  1502501d78  0bb492ef17
2           2   ADJAME  In Service  ...  NaN  0c2e182a0c  0bb492ef17
3           3  ASSINIE  In Service  ...  NaN  69bdfc69fd  8a1e6ce31f
4           4  ASSINIE  In Service  ...  NaN  7a1029d2ab  8a1e6ce31f

[5 rows x 12 columns]

In [ ]:
geo = pd.read_excel('cells_geo_description.xlsx')
geo

Column name                                        Description  Format
0         VILLES                                               City  String
1         STATUS                                  In Service or not  String
2   LOCALISATION                                  in ABIDJAN or not  String
3     DECOUPZONE                                  Geographical Zone  String
4       ZONENAME                                       Name of Zone  String
5      LONGITUDE                                          Longitude   Float
6       LATITUDE                                           Latitude   Float
7         REGION                                             Region  String
8           AREA                                               Area  String
9        CELL_ID                                     ID of the cell  String
10     SITE_CODE  Site (there are several cells per site, severa...  String

In [ ]:
cdr = pd.read_excel('CDR_description.xlsx')
cdr

Column name                                        Description   Format
0          PRODUCT                                       Voice or SMS   String
1            VALUE                                      Billing price  Integer
2        DATE_TIME               Time in format yyyy-MM-dd hh:mm:ss.0   String
3     CELL_ON_SITE  Which cell in the site was used (not needed here)  Integer
4  DW_A_NUMBER_INT  Anonymized phone number of the person for whic...   String
5  DW_B_NUMBER_INT        Anonymized phone number of the counterparty   String
6        COUNTRY_A                  Country of party A (useless here)   String
7        COUNTRY_B                  Country of party B (useless here)   String
8          CELL_ID                                     ID of the cell   String
9          SITE_ID                                     ID of the SITE   String

In [5]:
set1 = pd.read_csv('Telcom_dataset.csv')
set1.rename(columns={'PRODUTC': 'PRODUCT'}, inplace=True)
set1.head()

PRODUCT  VALUE               DATETIME  ... COUNTRY_B     CELL_ID     SITE_ID
0   Voice      0  2012-05-06 23:04:37.0  ...     19e2e  /TJNe+Mmtu  /TJNe+Mmtu
1     sms      0  2012-05-06 23:05:25.0  ...     19e2e  /+cKZKIp41  /+cKZKIp41
2     sms     25  2012-05-06 23:05:41.0  ...     19e2e  /u0FSD+ahi  /u0FSD+ahi
3   Voice     56  2012-05-06 23:05:42.0  ...     19e2e  +cNeJzsTp3  +cNeJzsTp3
4   Voice     14  2012-05-06 23:06:29.0  ...     19e2e  051KhYDCpv  051KhYDCpv

[5 rows x 10 columns]

In [4]:
set2 = pd.read_csv('Telcom_dataset2.csv')
set2.head()

PRODUCT  VALUE              DATE_TIME  ... COUNTRY_B     CELL_ID     SITE_ID
0     sms      0  2012-05-07 23:02:06.0  ...     19e2e  +854AcBQT2  +854AcBQT2
1   Voice      0  2012-05-07 23:03:44.0  ...     19e2e  +854AcBQT2  +854AcBQT2
2     sms      0  2012-05-07 23:04:06.0  ...     19e2e  +laSrk7g6q  +laSrk7g6q
3   Voice     37  2012-05-07 23:04:48.0  ...     19e2e  /xvRrCVKoQ  /xvRrCVKoQ
4     sms     25  2012-05-07 23:04:58.0  ...     19e2e  +bO+qkSonO  +bO+qkSonO

[5 rows x 10 columns]

In [7]:
set3 = pd.read_csv('Telcom_dataset3.csv')
set3.rename(columns= {"SIET_ID": "SITE_ID"}, inplace=True)
set3.head()

PRODUCT  VALUE              DATE_TIME  ... COUNTRY_B      CELLID     SITE_ID
0   Voice     61  2012-05-08 23:01:28.0  ...     19e2e  +38u2u/rfx  +38u2u/rfx
1     sms      0  2012-05-08 23:01:31.0  ...     19e2e  +Fs4mTvdKx  +Fs4mTvdKx
2   Voice     10  2012-05-08 23:02:59.0  ...     19e2e  +ZaaZozy/+  +ZaaZozy/+
3     sms      0  2012-05-08 23:03:04.0  ...     19e2e  +cNeJzsTp3  +cNeJzsTp3
4   Voice     22  2012-05-08 23:03:16.0  ...     19e2e  /eeMklgyJA  /eeMklgyJA

[5 rows x 10 columns]

The next step involves data exploration before cleaning and eventual analysis can be done.

### Possible problems to be addressed

In the cells dataframe, the 'Abidjan_EST' entry has unintended quotations from the source. Also, since each Telecom dataset represent specific dates, the date-time column needs to only represent time by itself. As such, it needs to be split and renamed. In the cells df, the Unamed column needs to be dropped as it serves no purpose.



## DATA CLEANING

### Uniformity

In [8]:
cells.rename(columns={'VILLES': 'City'}, inplace=True)
cells.replace(to_replace='"Abidjan_EST', value='Abidjan_EST', inplace=True)
cells.drop(['Unnamed: 0'], axis=1, inplace=True)
cells.head()

City      STATUS LOCALISATION  ... AREA     CELL_ID   SITE_CODE
0   ADJAME  In Service      ABIDJAN  ...  NaN  fbb2e75af3  0bb492ef17
1   ADJAME  In Service      ABIDJAN  ...  NaN  1502501d78  0bb492ef17
2   ADJAME  In Service      ABIDJAN  ...  NaN  0c2e182a0c  0bb492ef17
3  ASSINIE  In Service    INTERIEUR  ...  NaN  69bdfc69fd  8a1e6ce31f
4  ASSINIE  In Service    INTERIEUR  ...  NaN  7a1029d2ab  8a1e6ce31f

[5 rows x 11 columns]

Eliminating redundant columns in each dataset

In [9]:
eliminate = ['CELL_ON_SITE', 'COUNTRY_A', 'COUNTRY_B']
set1.drop(eliminate, axis=1,inplace=True)
set2.drop(eliminate, axis=1, inplace=True)
set3.drop(eliminate, axis=1, inplace=True)

### Completeness

In [10]:
cells.isnull().sum().sort_values(ascending=False)

STATUS          67
AREA            23
ZONENAME         6
SITE_CODE        0
CELL_ID          0
REGION           0
LATITUDE         0
LONGITUDE        0
DECOUPZONE       0
LOCALISATION     0
City             0
dtype: int64

In [14]:
cells.dropna(inplace=True)

In [11]:
set1.isnull().sum().sort_values(ascending=False)

SITE_ID            716
CELL_ID              0
DW_B_NUMBER_INT      0
DW_A_NUMBER_INT      0
DATETIME             0
VALUE                0
PRODUCT              0
dtype: int64

In [15]:
set1.dropna(inplace=True)

In [12]:
set2.isnull().sum().sort_values(ascending=False)

SITE_ID        676
CELL_ID          0
DW_B_NUMBER      0
DW_A_NUMBER      0
DATE_TIME        0
VALUE            0
PRODUCT          0
dtype: int64

In [16]:
set2.dropna(inplace=True)

In [13]:
set3.isnull().sum().sort_values(ascending=False)

SITE_ID            605
CELLID               0
DW_B_NUMBER_INT      0
DW_A_NUMBER_INT      0
DATE_TIME            0
VALUE                0
PRODUCT              0
dtype: int64

In [17]:
set3.dropna(inplace=True)

### Consistency

In [21]:
cells.duplicated().sum()

17

In [22]:
set1.duplicated().sum()

70

In [23]:
set2.duplicated().sum()

62

In [25]:
set3.duplicated().sum()

85

In [26]:
# Dropping Duplicates in datasets

cells.drop_duplicates(inplace=True)
set1.drop_duplicates(inplace=True)
set2.drop_duplicates(inplace=True)
set3.drop_duplicates(inplace=True)

### Merging Datasets

In [29]:
# Merging the three days' datasets

days = pd.concat([set1,set2,set3], axis=0, join='outer', ignore_index=True)
days.head()

PRODUCT  VALUE               DATETIME  ... DW_A_NUMBER DW_B_NUMBER CELLID
0   Voice      0  2012-05-06 23:04:37.0  ...         NaN         NaN    NaN
1     sms      0  2012-05-06 23:05:25.0  ...         NaN         NaN    NaN
2     sms     25  2012-05-06 23:05:41.0  ...         NaN         NaN    NaN
3   Voice     56  2012-05-06 23:05:42.0  ...         NaN         NaN    NaN
4   Voice     14  2012-05-06 23:06:29.0  ...         NaN         NaN    NaN

[5 rows x 11 columns]

Enhancing the new concatated table

In [30]:
drop = ['DATE_TIME', 'CELLID', 'DW_A_NUMBER', 'DW_B_NUMBER']
days.drop(drop, axis=1, inplace=True)
days.head()

PRODUCT  VALUE  ...     CELL_ID     SITE_ID
0   Voice      0  ...  /TJNe+Mmtu  /TJNe+Mmtu
1     sms      0  ...  /+cKZKIp41  /+cKZKIp41
2     sms     25  ...  /u0FSD+ahi  /u0FSD+ahi
3   Voice     56  ...  +cNeJzsTp3  +cNeJzsTp3
4   Voice     14  ...  051KhYDCpv  051KhYDCpv

[5 rows x 7 columns]

Merging all datasets now and cleaning them

In [31]:
df = pd.merge(days, cells, how='inner', left_on=['SITE_ID'], right_on=['SITE_CODE'])
df.head()

PRODUCT  VALUE               DATETIME  ...          AREA   CELL_ID_y   SITE_CODE
0     sms      0  2012-05-07 00:00:00.0  ...  CENTRE NORTH  24ccdda3d5  1b5540c02d
1     sms      0  2012-05-07 00:00:00.0  ...  CENTRE NORTH  ebfc58af70  1b5540c02d
2     sms      0  2012-05-07 00:00:00.0  ...  CENTRE NORTH  ba8ce6e4ef  1b5540c02d
3     sms      0  2012-05-07 00:00:00.0  ...  CENTRE NORTH  c2545171dc  1b5540c02d
4     sms      0  2012-05-07 00:00:00.0  ...  CENTRE NORTH  f15bd5483f  1b5540c02d

[5 rows x 18 columns]

In [32]:
# Dropping redundant columns

eliminate = ['DW_B_NUMBER_INT', 'CELL_ID_x', 'CELL_ID_y', 'SITE_CODE']
df.drop(eliminate, axis=1, inplace=True)
df.head()

PRODUCT  VALUE  ...             REGION          AREA
0     sms      0  ...  VALLEE DU BANDAMA  CENTRE NORTH
1     sms      0  ...  VALLEE DU BANDAMA  CENTRE NORTH
2     sms      0  ...  VALLEE DU BANDAMA  CENTRE NORTH
3     sms      0  ...  VALLEE DU BANDAMA  CENTRE NORTH
4     sms      0  ...  VALLEE DU BANDAMA  CENTRE NORTH

[5 rows x 14 columns]

## ANALYSIS

In [33]:
# Regions with the highest number of calls

df['REGION'].value_counts().to_frame()

REGION
LAGUNE              31953
LACS                 2408
BAS SASSANDRA        2327
HAUT SASSANDRA       2274
VALLEE DU BANDAMA    1551
SUD BANDAMA          1386
FROMAGER              904
SAVANE                895
WORODOUGOU            792
18 MONTAGNES          687
MARAHOUE              618
AGNEBY                579
NZI COMOE             563
MOYEN CAVALLY         528
DENGUELE              395
ZANZAN                315
MOYEN COMOE           315
SUD COMOE             198
BAFING                 99

Taking the top 10 regions with highest traffic:
* LAGUNE               31953
* LACS                  2408
* BAS SASSANDRA         2327
* HAUT SASSANDRA        2274
* VALLEE DU BANDAMA     1551
* SUD BANDAMA           1386
* FROMAGER              904
* SAVANE                895
* WORODOUGOU            792
* 18 MONTAGNES          687


In [40]:
high_regions =['LAGUNE', 'LACS', 'BAS SASSANDRA', 'HAUT SASSANDRA', 'VALLEE DU BANDAMA',
              'SUD BANDAMA', 'SAVANE', 'FROMAGER' 'WORODOUGOU' ,'MONTAGNES']

hr_df = df[df['REGION'].isin(high_regions)]
hr_df.head()

PRODUCT  VALUE  ...             REGION          AREA
0     sms      0  ...  VALLEE DU BANDAMA  CENTRE NORTH
1     sms      0  ...  VALLEE DU BANDAMA  CENTRE NORTH
2     sms      0  ...  VALLEE DU BANDAMA  CENTRE NORTH
3     sms      0  ...  VALLEE DU BANDAMA  CENTRE NORTH
4     sms      0  ...  VALLEE DU BANDAMA  CENTRE NORTH

[5 rows x 14 columns]

In [41]:
# Which Cities are in high traffic zones?

df['ZONENAME'].value_counts()

ZONE C - Abidjan_OUEST     10121
ZONE E - Abidjan_NORD       6626
ZONE B - Grand-SUD          5411
ZONE B - Abidjan_EST        4760
ZONE D - Abidjan_SUD        4602
ZONE D - Grand-OUEST        4233
ZONE A - Abidjan_CENTRE     4121
ZONE A - Grand-NORD         3843
ZONE E - SUD-OUEST          2993
ZONE C - Grand-EST          1745
ZONE D - Abidjan_EST         186
ZONE C - Abidjan_EST          27
ZONE A - Abidjan_OUEST        21
ZONE A - Grand-SUD            18
ZONE B - Grand-EST            18
ZONE B - Abidjan_OUEST        18
ZONE C - Abidjan_SUD          15
ZONE A - Grand-EST             6
ZONE B - Abidjan_SUD           6
ZONE D - Grand-SUD             6
ZONE B - Grand-NORD            3
ZONE B - Abidjan-EST           3
ZONE B - Abidjan_Nord          3
ZONE B - Abidjan_CENTRE        2
Name: ZONENAME, dtype: int64

The top 10 cities in the high traffic zones are:
* ZONE C - Abidjan_OUEST     10121
* ZONE E - Abidjan_NORD       6626
* ZONE B - Grand-SUD          5411
* ZONE B - Abidjan_EST        4760
* ZONE D - Abidjan_SUD        4602
* ZONE D - Grand-OUEST        4233
* ZONE A - Abidjan_CENTRE     4121
* ZONE A - Grand-NORD         3843
* ZONE E - SUD-OUEST          2993
* ZONE C - Grand-EST          1745

In [42]:
high_zones = ['ZONE C - Abidjan_OUEST','ZONE E - Abidjan_NORD', 'ZONE B - Grand-SUD',
              'ZONE B - Abidjan_EST', 'ZONE D - Abidjan_SUD', 'ZONE A - Grand-NORD',
              'ZONE E - SUD-OUEST', 'ZONE C - Grand-EST', 'ZONE D - Abidjan_EST']
hz_df = df[df['ZONENAME'].isin(high_zones)]
hz_df.head()

PRODUCT  VALUE  ...             REGION          AREA
0     sms      0  ...  VALLEE DU BANDAMA  CENTRE NORTH
1     sms      0  ...  VALLEE DU BANDAMA  CENTRE NORTH
2     sms      0  ...  VALLEE DU BANDAMA  CENTRE NORTH
3     sms      0  ...  VALLEE DU BANDAMA  CENTRE NORTH
4     sms      0  ...  VALLEE DU BANDAMA  CENTRE NORTH

[5 rows x 14 columns]

In [43]:
# What is product usage like in these high traffic zones?

usage = hz_df['PRODUCT'].value_counts().to_frame()
usage

PRODUCT
sms      23505
Voice    16782

In [44]:
#Which of the two is the more profitable product?

profit = hz_df.groupby(['PRODUCT'])['VALUE'].sum()
(profit / profit.sum()) * 100

PRODUCT
Voice    83.390392
sms      16.609608
Name: VALUE, dtype: float64

In [45]:
# Most profitable zone is?

pd.pivot_table(data=hz_df, values='VALUE', index=['PRODUCT', 'ZONENAME'], aggfunc=np.sum)

VALUE
PRODUCT ZONENAME                      
Voice   ZONE A - Grand-NORD     101088
        ZONE B - Abidjan_EST    182138
        ZONE B - Grand-SUD       99114
        ZONE C - Abidjan_OUEST  171186
        ZONE C - Grand-EST       78725
        ZONE D - Abidjan_EST      8196
        ZONE D - Abidjan_SUD     99242
        ZONE E - Abidjan_NORD   108496
        ZONE E - SUD-OUEST       84489
sms     ZONE A - Grand-NORD      13073
        ZONE B - Abidjan_EST     25147
        ZONE B - Grand-SUD       23184
        ZONE C - Abidjan_OUEST   50578
        ZONE C - Grand-EST        7643
        ZONE D - Abidjan_EST       750
        ZONE D - Abidjan_SUD     24312
        ZONE E - Abidjan_NORD    30851
        ZONE E - SUD-OUEST       10231

In [60]:
# Profits in High Trafffic Zones

profit_hz = hz_df.groupby(['PRODUCT'])['VALUE'].sum()

(profit_hz/ profit_hz.sum() * 100).to_frame()

VALUE
PRODUCT           
Voice    83.390392
sms      16.609608

Most profitable product in high traffic zones is the Voice call.

In [51]:
pd.pivot_table(data=hz_df, values='VALUE', index=['City'], aggfunc=np.sum).sort_values(by='VALUE', ascending=False)

VALUE
City                 
YOPOUGON       198556
COCODY         169678
ABOBO          119236
KOUMASSI        54958
ADJAME          45482
...               ...
TIEMELEKRO          0
SONGON              0
NGATTADOLIKRO       0
BROFODOUME          0
SAPLI               0

[301 rows x 1 columns]

In [55]:
# Regions with highest frequency of calls

calls = df[df['PRODUCT']== 'Voice']


# Grouping by Region and aggregating by value of calls

calls.groupby(['REGION'])['VALUE'].sum().to_frame().sort_values(by='VALUE', ascending=False)

VALUE
REGION                   
LAGUNE             711553
BAS SASSANDRA       61076
LACS                45489
HAUT SASSANDRA      42275
VALLEE DU BANDAMA   39072
NZI COMOE           29357
SUD BANDAMA         27456
FROMAGER            25661
18 MONTAGNES        25299
SAVANE              23241
WORODOUGOU          20976
MOYEN CAVALLY       18828
MARAHOUE            14085
SUD COMOE           12180
DENGUELE            12180
AGNEBY              11316
ZANZAN              10611
MOYEN COMOE         10533
BAFING               3411

Regions with the highest frequency of calls



1.   LAGUNE                                711553
2.   BAS SASSANDRA                         61076
3.   LAC S                                 45489
4.   HAUT SASSANDRA                        42275
5.   VALLEE DU BANDAMA                     39072
6.   NZI COMOE                             29357



In [64]:
# Most profitable cities the company has invested in

city_profits = df.groupby(['City'])[['VALUE']].sum()
city_profits.sort_values(by='VALUE', ascending=False)

VALUE
City                
YOPOUGON      198556
COCODY        177745
ABOBO         119236
KOUMASSI       71356
TREICHVILLE    65557
...              ...
PAKAOUBO           0
OUELLE             0
SIENSOBA           0
TORO-KINKENE       0
BROFODOUME         0

[398 rows x 1 columns]

## CONCLUSIONS

* The regions with the highest frequency of calls: Lagune, Bas Sassandra, Lac S, Haut Sassandra, Vallee du Bandama.

* The cities inside high traffic zones are: Yopougon, Cocody, Abobo, Koumassi and Adjame.

* The most profitable product offered by the company is the Voice call which accounts for 83% of the profit. This is also consistent in high traffic zones

* Overall, the most profitable cities the company has invested in are: Yopougon, Cocody, Abobo, Koumassi and Treichville.